app/xvctl/xvctl.1
.\"	$OpenBSD: xvctl.1,v 1.3 2013/08/26 19:25:10 espie Exp $
.\"
.\" Copyright (c) 2001 Marc Espie
.\"
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE DEVELOPERS ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
.\" OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
.\" IN NO EVENT SHALL THE DEVELOPERS BE LIABLE FOR ANY DIRECT, INDIRECT,
.\" INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
.\" NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
.\" DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
.\" THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
.\" (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
.\" THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd November 9, 2001
.Dt XVCTL 1
.Os
.Sh NAME
.Nm xvctl
.Nd get or set xv extension values
.Sh SYNOPSIS
.Nm xvctl
.Op Fl n
.Op Fl d Ar display
.Op Fl s Ar screen
.Op Fl A Ar adaptor
.Ar name ...
.Nm xvctl
.Op Fl n
.Op Fl d Ar display
.Op Fl s Ar screen
.Op Fl A Ar adaptor
.Fl w
.Ar name=value ...
.Nm xvctl
.Op Fl n
.Op Fl d Ar display
.Op Fl s Ar screen
.Op Fl A Ar adaptor
.Fl a
.Sh DESCRIPTION
The
.Nm
utility retrieves the Xvideo extension attributes and
allows the user to set them.
.Pp
For instance, this can be used to change the contrast or brightness of an
external video program.
.Pp
The options are as follows:
.Bl -tag -width xxx
.It Fl d Ar display
affect the display named
.Ar display , 
instead of the default display.
.It Fl s Ar screen
affect screen #
.Ar screen .
By default, the first screen with video adaptors will be chosen.
.It Fl A Ar adaptor
affect adaptor #
.Ar adaptor .
By default, the first display adaptor will be chosen.
.It Fl a 
list all the currently available gettable attributes.
This is the default, if no parameters are givne.
.It Fl n
suppress printing of the field name, only output the field value.
Useful for setting shell variables.
.El
.Sh SEE ALSO
.Xr xvinfo 1 ,
.Xr Xv 3x


xserver/ChangeLog
commit e597f0119cd69b6d9edf86d06d941468f90d8e6d
Author: Jeremy Huddleston <jeremyhu@apple.com>
Date:   Fri Jan 27 21:07:17 2012 -0800

    configure.ac: 1.11.4
    
    Signed-off-by: Jeremy Huddleston <jeremyhu@apple.com>

commit 6c491463ee7a3637c8aae7ddae9625b36b792078
Author: Jeremy Huddleston <jeremyhu@apple.com>
Date:   Fri Jan 20 23:54:58 2012 -0800

    1.11.3.902 (1.11.4 RC2)
    
    Signed-off-by: Jeremy Huddleston <jeremyhu@apple.com>

commit b747a91485184649c05aba6fa10734f70e926032
Author: Jeremy Huddleston <jeremyhu@apple.com>
Date:   Wed Jan 18 11:47:21 2012 -0800

    XQuartz: Bump bundle version to 2.7.1
    
    Signed-off-by: Jeremy Huddleston <jeremyhu@apple.com>

commit a1ed67fa0551ee2b39d002d3b786a2017b37c7ac
Author: Peter Hutterer <peter.hutterer@who-t.net>
Date:   Fri Jan 6 13:20:45 2012 +1000

    render: don't bother with animated cursors on floating slaves (#39989)
    
    X.Org Bug 39989 <http://bugs.freedesktop.org/show_bug.cgi?id=39989>
    
    Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>
    Reviewed-by: Jeremy Huddleston <jeremyhu@apple.com>
    (cherry picked from commit bbb6b8c834e0e1491ca14403b5d0840dd14380d3)

commit f89c39e6c9e20ffc3c1ae228ee601c2ab9007ff4
Author: Peter Hutterer <peter.hutterer@who-t.net>
Date:   Mon Jan 9 15:04:10 2012 +1000

    dix: Update pointer limits for floating devices too (#43635)
    
    When the screen is restructured, the pointer limits need to be reset for
    floating slave devices as well, not just for master pointers. Only skip
    devices that don't have a cursor (attached slaves and keyboard)
    
    Bug reproducer: float an absolute slave device, rotate the screen - the
    device is now confined to a section of the screen only.
    
    X.Org Bug 43635 <http://bugs.freedesktop.org/show_bug.cgi?id=43635>
    
    Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>
    Reviewed-by: Chase Douglas <chase.douglas@canonical.com>
    (cherry picked from commit 35bd77e9d0701daae87d681900d749604fc6471f)

commit dea9377a257a402b2c7fe4a96c958d951a7d4579
Author: Zhigang Gong <zhigang.gong@linux.intel.com>
Date:   Wed Jan 4 07:01:19 2012 +0000

    mi/mibitblt: Fix an overflow bug of bit shift.
    
    When depth equal to 32 and planeMask equal to 0, the overflow will
    occur and cause the pixmap can't be cleared. There are some test
    cases in XTS hit this bug, and this fix can eliminate the corresponding
    failures.
    
    Signed-off-by: Zhigang Gong <zhigang.gong@linux.intel.com>
    Reviewed-by: Adam Jackson <ajax@redhat.com>
    Signed-off-by: Keith Packard <keithp@keithp.com>
    (cherry picked from commit f4956faab9ccf9aba6cf9603f4489f5dad19a347)

commit d2db5100b1bafeeea8c1febd390631c8f8abb094
Author: Alan Coopersmith <alan.coopersmith@oracle.com>
Date:   Fri Oct 28 21:29:50 2011 -0700

    Convert DetermineClientCmd to use strdup instead of malloc+strncpy
    
    *cmdname is initialized to NULL earlier in the function, so it's
    okay to overwrite it with NULL if strdup fails, don't need that
    extra check.
    
    Signed-off-by: Alan Coopersmith <alan.coopersmith@oracle.com>
    Reviewed-by: Jeremy Huddleston <jeremyhu@apple.com>
    (cherry picked from commit 780133f9ae7fada462714b47e79d26075bbd9abe)

commit 5b28bcb340e279e7f4e76617e86b439092367690
Author: Adam Jackson <ajax@redhat.com>
Date:   Tue Dec 13 21:23:12 2011 -0500

    dix: Tune dixLookupDrawable for success
    
    The vast vast vast majority of resource lookups are successful.  Move some
    work to the error paths so we don't punish success.
    
    Before:
    40000000 trep @   0.0009 msec (1109091.3/sec): PutImage 10x10 square
    60000000 trep @   0.0005 msec (2072652.2/sec): ShmPutImage 10x10 square
    
    After:
    40000000 trep @   0.0009 msec (1148346.9/sec): PutImage 10x10 square
    60000000 trep @   0.0005 msec (2091666.1/sec): ShmPutImage 10x10 square
    
    Reviewed-by: Alan Coopersmith <alan.coopersmith@oracle.com>
    Signed-off-by: Adam Jackson <ajax@redhat.com>
    (cherry picked from commit 8f9bdfd293ad8e45755efe8d764b4dcc2a724f51)

commit e1ea93fec93b13d18df4451210aaaf322b769b6a
Author: Adam Jackson <ajax@redhat.com>
Date:   Tue Dec 13 20:23:40 2011 -0500

    dix: Tune dtrace hooks around Dispatch
    
    Don't call LookupMajorName if the hooks aren't active, it's quite expensive.
    
    Before:
    40000000 trep @   0.0009 msec (1087458.5/sec): PutImage 10x10 square
    60000000 trep @   0.0005 msec (2012238.6/sec): ShmPutImage 10x10 square
    
    After:
    40000000 trep @   0.0009 msec (1109091.3/sec): PutImage 10x10 square
    60000000 trep @   0.0005 msec (2072652.2/sec): ShmPutImage 10x10 square
    
    Reviewed-by: Alan Coopersmith <alan.coopersmith@oracle.com>
    Signed-off-by: Adam Jackson <ajax@redhat.com>
    (cherry picked from commit 83a98543b58c661a22574a6f8d8f9d777c0955b8)

commit 9eaa0b5c89f645c50ea90e07e8b49ebacf4349a0
Author: Adam Jackson <ajax@redhat.com>
Date:   Tue Dec 13 21:00:05 2011 -0500

    xace: ricer tuning for XaceHook
    
    gcc doesn't want to hoist the check for XaceHooks[hook] != NULL above the
    varargs code for some reason, so do it ourselves.
    
    Before:
    40000000 trep @   0.0010 msec (1050420.2/sec): PutImage 10x10 square
    60000000 trep @   0.0005 msec (1921147.6/sec): ShmPutImage 10x10 square
    
    After:
    40000000 trep @   0.0009 msec (1087458.5/sec): PutImage 10x10 square
    60000000 trep @   0.0005 msec (2012238.6/sec): ShmPutImage 10x10 square
    
    Reviewed-by: Alan Coopersmith <alan.coopersmith@oracle.com>
    Signed-off-by: Adam Jackson <ajax@redhat.com>
    (cherry picked from commit 11977fab546da35d579ebe285e3c26864007805e)

commit a3b6ecc893e17028cfbe16d8d7981c9037fc0766
Author: Peter Hutterer <peter.hutterer@who-t.net>
Date:   Tue Jan 3 10:45:45 2012 +1000

    xfree86: split warning about missing identifier or input driver
    
    Check for identifier first and bail if it's missing (also remove the current

app/xvctl/xvctl.c
/*	$OpenBSD: xvctl.c,v 1.3 2013/08/26 19:25:10 espie Exp $	*/
/*
 * Copyright (c) 2001 Marc Espie.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE OPENBSD PROJECT AND CONTRIBUTORS
 * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE OPENBSD
 * PROJECT OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
#include <X11/X.h>
#include <X11/Xlib.h>
#include <X11/extensions/Xvlib.h>
#include <unistd.h>
#include <stdlib.h>
#include <stdio.h>
#include <err.h>
#include <string.h>
#include <ctype.h>

void
usage()
{
	fprintf(stderr,  "usage:\t%s\n\t%s\n\t%s\n",
	    "xvctl [-n] [-d display] [-s screen] [-A adaptor] variable ...",
	    "xvctl [-n] [-d display] [-s screen] [-A adaptor] variable=value ...",
	    "xvctl [-n] [-d display] [-s screen] [-A adaptor] -a");
	exit(1);
}

char *
reformat(char *name)
{
	static char result[BUFSIZ];
	char *s;
	char *t;

	if (strlcpy(result, name, sizeof result) >= sizeof result)
		warnx("Name %s was truncated", name);
	s = result;
	if (strncmp(result, "XV_", 3) == 0)
		s+=3;
	for (t = s; *t; t++)
		*t = tolower(*t);
    	return s;
}

XvAttribute *
findname(XvAttribute table[], int nattr, const char *name)
{
	int i;

	if (strncasecmp(name, "XV_", 3) == 0)
		name +=3; 
	for (i = 0; i < nattr; i++) {
		if (strcasecmp(table[i].name+3, name) == 0)
			return table+i;
	}
	warnx("No such attribute %s", name);
	return NULL;
}

void
printattribute(Display *dpy, XvPortID id, int nflag, XvAttribute *a)
{
	Atom atom;
	int v;

	atom = XInternAtom(dpy, a->name, True);
	if (atom != None)
		if (Success == XvGetPortAttribute(dpy, id, atom, &v)) {
			if (!nflag)
				printf("%s=", reformat(a->name));
			printf("%d\n", v);
		}
}

void
showall(Display *dpy, XvPortID id, int nflag, XvAttribute table[], int nattr)
{
	int i;

	for (i = 0; i < nattr; i++)
		if (table[i].flags & XvGettable)
			printattribute(dpy, id, nflag, table+i);
}

void
setvar(Display *dpy, XvPortID id, int nflag, XvAttribute table[], int nattr,
    const char *name, int v)
{
	XvAttribute *at;
	Atom atom;

	at = findname(table, nattr, name);
	if (!at) 
		return;
	if (!(at->flags & XvSettable)) {
		warnx("Can't set %s", name);
		return;
	}
	if ((at->max_value != -1 && v > at->max_value)  || v < at->min_value) {
		warnx("Value %d for %s out of range (%d-%d)", v, name,
		    at->min_value, at->max_value);
		return;
	}
	atom = XInternAtom(dpy, at->name, True);
	if (atom != None)
		if (Success == XvSetPortAttribute(dpy, id, atom, v)) {
			if (!nflag)
				printattribute(dpy, id, nflag, at);
		}
}

void
showvar(Display *dpy, XvPortID id, int nflag, XvAttribute table[], int nattr,
    const char *name)
{
	XvAttribute *at;

	at = findname(table, nattr, name);
	if (!at)
		return;
    	if (!(at->flags & XvGettable)) {
		warnx("Can't get %s", name);
		return;
	}
	printattribute(dpy, id, nflag, at);
}

void
parse(Display *dpy, XvPortID id, int nflag, XvAttribute table[],
    int nattr, const char *s)
{
	char buf[BUFSIZ];
	char *cp;

	if (strlcpy(buf, s, sizeof buf) >= sizeof(buf))
		warnx("Buffer too small for %s", s);
	cp = strchr(buf, '=');
	if (cp != NULL) {
		long v;
		char *end;
		char *tail;

		tail = cp;
		while (tail > buf && isspace(tail[-1]))
			tail--;
		*tail = 0;
		do {
			cp++;
		} while (isspace(*cp));
		v = strtol(cp, &end, 0);
		if (*end != 0) {
			warnx("Bogus value: %s", cp);
			return;
		}
		setvar(dpy, id, nflag, table, nattr, buf, v);
	} else
		showvar(dpy, id, nflag, table, nattr, buf);
}

char *
displayname(char *display)
{
	return display ? display : XDisplayName(display);
}

int
main(int argc, char *argv[])
{
	int ch;
	int aflag = 0;
	int nflag = 0;
	char *display = NULL;
	int screen = -1;
	int nscreens;
	int nattr;
	unsigned int nadaptors;
	XvAdaptorInfo *ainfo;
	XvAttribute *table;
	int adaptor = 0;
	Display *dpy;
	unsigned int ver, rev, eventB, reqB, errorB;

	while ((ch = getopt(argc, argv, "ad:ns:A:w")) != -1) {
		switch(ch) {
		case 'a':
			aflag = 1;
			break;
		case 'd':
			display = optarg;
			break;
		case 's':
			screen = atoi(optarg);
			break;
		case 'A':
			adaptor = atoi(optarg);
			break;
		case 'n':
			nflag = 1;
			break;
		case 'w':
			/* flag no longer needed */
			break;
		default:
			usage();
		}
	}
	argc -= optind;
	argv += optind;

	dpy = XOpenDisplay(display);
	if (!dpy)
		errx(1, "Unable to open display %s", displayname(display));
	if (Success != XvQueryExtension(dpy, &ver, &rev, &reqB, &eventB,
	    &errorB)) 
	    	errx(1, "No X-Video extension on %s", displayname(display));

	if (!nflag)
		printf("# X-Video extension version %i.%i on %s\n", ver, rev,
		    displayname(display));

	nscreens = ScreenCount(dpy);
	if (screen == -1) {
		for (screen = 0; screen < nscreens; screen++) {
			XvQueryAdaptors(dpy, RootWindow(dpy, screen),
			    &nadaptors, &ainfo);
			if (nadaptors)
				break;
		}
	}
	if (screen >= nscreens)
		errx(1, "Not a screen with adaptors");

	if (adaptor > nadaptors)
		errx(1, "No such adaptor");

	table = XvQueryPortAttributes(dpy, ainfo[adaptor].base_id, &nattr);


	if (argc == 0) {
		showall(dpy, ainfo[adaptor].base_id, nflag, table, nattr);
	}
	for (; *argv != NULL; ++argv)
		parse(dpy, ainfo[adaptor].base_id, nflag, table,
		    nattr, *argv);
	XFree(table);
	exit(0);
}
